lib/Fmt.ml

```diff
@@ -124,16 +124,17 @@ let list_pn x1N pp =
   | [] -> noop
   | [x1] -> lazy_ (fun () -> pp ~prev:None x1 ~next:None)
   | x1 :: (x2 :: _ as x2N) ->
-      lazy_ (fun () -> pp ~prev:None x1 ~next:(Some x2))
-      $
-      let rec list_pn_ prev = function
-        | [] -> noop
-        | [xI] -> lazy_ (fun () -> pp ~prev:(Some prev) xI ~next:None)
-        | xI :: (xJ :: _ as xJN) ->
-            lazy_ (fun () -> pp ~prev:(Some prev) xI ~next:(Some xJ))
-            $ list_pn_ xI xJN
+      let l =
+        let rec aux (prev, acc) = function
+          | [] -> List.rev acc
+          | [xI] -> aux (xI, (Some prev, xI, None) :: acc) []
+          | xI :: (xJ :: _ as xJN) ->
+              aux (xI, (Some prev, xI, Some xJ) :: acc) xJN
+        in
+        aux (x1, [(None, x1, Some x2)]) x2N
       in
-      list_pn_ x1 x2N
+      with_pp (fun fs ->
+          List.iter l ~f:(fun (prev, x, next) -> eval fs (pp ~prev x ~next)) )
```

Comment:
I did something like this at first but it generated an error on some other tests (comments.ml I think, the comments were not printed in the correct order), I hope if something was wrong with this implementation it would show on the tests or with test_branch, the issue with the minimal interface is that it made it difficult to write list_pn in a tail recursive way

Reply:
The thing's that's happening is that side effects were not properly delayed. I'll write a unit test for list_pn instead of having formatting tests fail for that.

The following works:

```ocaml
List.fold_left l ~init:noop ~f:(fun acc (prev, x, next) ->
    acc $ lazy_ (fun () -> pp ~prev x ~next) )
```

Reply:
Indeed it's better, thanks! +1 for some unit tests